### PR TITLE
feat(tensorflow backend): manual dtype casting removal

### PR DESCRIPTION
**Pull request by Madjid-CH**
# PR Description
this PR is for this [task](https://trello.com/c/kSyDbng5)
I came ready this time 😃 I used this script to figure out which dtypes are supported and which are not.
```python
import tensorflow as tf


dtypes = [
    tf.bfloat16,
    tf.float16,
    tf.float32,
    tf.float64,
    tf.int8,
    tf.int16,
    tf.int32,
    tf.int64,
    tf.uint8,
    tf.uint16,
    tf.uint32,
    tf.uint64,
    tf.bool,
    tf.complex64,
    tf.complex128,
    tf.double,
    tf.half,
]


def get_supported_and_unsupported_dtypes():
    supported_dtypes = set()
    unsupported_dtypes = set()

    for dtype in dtypes:
        try:
            x = tf.constant([[1, 2,] , [2, 1]], dtype=dtype)
            x1 = tf.constant([1, 2, 3], dtype=dtype)
            x2 = tf.constant([1, 2, 3], dtype=dtype)
            print(cummin(x))
            supported_dtypes.add(dtype)
        except Exception as e:
            print(e)
            unsupported_dtypes.add(dtype)

    return supported_dtypes, unsupported_dtypes


supported_dtypes, unsupported_dtypes = get_supported_and_unsupported_dtypes()

supported_dtypes = [repr(d).replace("tf.", "") for d in supported_dtypes]
unsupported_dtypes = [repr(d).replace("tf.", "") for d in unsupported_dtypes]

print("Supported Data Types:")
m = f"""@with_supported_dtypes(
        {{"2.13.0 and below": {tuple(supported_dtypes)}}}, backend_version
    )"""
print(m)

print("\nUnsupported Data Types:")
m = f"""@with_unsupported_dtypes(
        {{"2.13.0 and below": {tuple(unsupported_dtypes)}}}, backend_version
    )"""
print(m)

```

### CODE REVIEW

File: ivy/functional/backends/tensorflow/elementwise.py [hunk 1]
```diff
@@ -600,10 +596,7 @@ def positive(
     return tf.experimental.numpy.positive(x)
 
 
-@with_unsupported_dtypes(
-    {"2.13.0 and below": ("uint8", "uint16", "uint32", "uint64", "float64")},
-    backend_version,
-)
+@with_unsupported_dtypes({"2.13.0 and below": ("bool", "unsigned")}, backend_version)
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/elementwise.py:599]: have you tested `float64`?

File: ivy/functional/backends/tensorflow/elementwise.py [hunk 2]
```diff
@@ -812,18 +792,7 @@ def erf(
...
-            "uint8",
-            "uint16",
-            "uint32",
-            "uint64",
-            "complex",
-        )
-    },
-    backend_version,
-)
+@with_unsupported_dtypes({"2.13.0 and below": ("complex", "bool")}, backend_version)
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/elementwise.py:795]: have you tested uints for this function?

File: ivy/functional/backends/tensorflow/elementwise.py [hunk 3]
```diff
@@ -833,26 +802,10 @@ def maximum(
...
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/elementwise.py:808]: same as `maximum`

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 4]
```diff
@@ -663,12 +667,8 @@ def cummax(
     dtype: Optional[tf.DType] = None,
     out: Optional[Union[tf.Tensor, tf.Variable]] = None,
 ) -> Tuple[tf.Tensor, tf.Tensor]:
-    if x.dtype in (tf.bool, tf.float16):
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/experimental/statistical.py:666]: do all these `dtypes` work without any type casting?

File: ivy/functional/backends/tensorflow/linear_algebra.py [hunk 5]
```diff
@@ -587,10 +589,6 @@ def tensordot(
 ) -> Union[tf.Tensor, tf.Variable]:
     # find type to promote to
     dtype = ivy.as_native_dtype(ivy.promote_types(x1.dtype, x2.dtype))
-
-    # type casting to float32 which is acceptable for tf.tensordot
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/linear_algebra.py:591]: seeing this comment, just wanted to confirm if you've tested with all dtypes other than `float32`

File: ivy/functional/backends/tensorflow/linear_algebra.py [hunk 6]
```diff
@@ -615,8 +613,7 @@ def trace(
 
 
 @with_unsupported_dtypes(
-    {"2.13.0 and below": ("bfloat16", "float16", "complex")},
-    backend_version,
+    {"2.13.0 and below": ("int16", "int8", "bool", "unsigned")}, backend_version
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/linear_algebra.py:616]: do `bfloat16` and `float16` work for the function?

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Hey @Madjid-CH, I know you've already run a script for this, but just wanted to ask if you could double-check these specific instances? I think there could be a reason why these were set as they were before so would be great to be 100% sure before changing. Could you please also run the tests for these functions to see if the changes have any impact on them? Thanks 😄

File: ivy/functional/backends/tensorflow/elementwise.py (hunk 1 above)
Comment (Madjid-CH) [reply ivy/functional/backends/tensorflow/elementwise.py:599]: yes, I rerun it and `float64` does not cause an issue
image

File: ivy/functional/backends/tensorflow/elementwise.py (hunk 2 above)
Comment (Madjid-CH) [reply ivy/functional/backends/tensorflow/elementwise.py:795]: yep:
image

File: ivy/functional/backends/tensorflow/elementwise.py (hunk 3 above)
Comment (Madjid-CH) [reply ivy/functional/backends/tensorflow/elementwise.py:808]:
```
for d in ["uint8","uint16","uint32","uint64", ]:
    x1 = tf.constant([1, 2, 3], dtype=d)
    x2 = tf.constant([1, 2, 3], dtype=d)
    print(minimum(x1, x2))
    
tf.Tensor([1 2 3], shape=(3,), dtype=uint8)
tf.Tensor([1 2 3], shape=(3,), dtype=uint16)
tf.Tensor([1 2 3], shape=(3,), dtype=uint32)
tf.Tensor([1 2 3], shape=(3,), dtype=uint64)
```

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 4 above)
Comment (Madjid-CH) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:666]: for d in ('float32', 'float64', 'int32', 'uint8', 'int16', 
          'int8', 'complex64', 'int64', 'bfloat16', 'uint16', 
          'complex128', 'float16', 'uint32', 'uint64'): 
    x = tf.constant([1, 2,], dtype=d)
    print(cummax(x))        
    
(<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 2.], dtype=float32)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float64, numpy=array([1., 2.])>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int32, numpy=array([1, 2])>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint8, numpy=array([1, 2], dtype=uint8)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int16, numpy=array([1, 2], dtype=int16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int8, numpy=array([1, 2], dtype=int8)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 2.], dtype=float32)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int64, numpy=array([1, 2], dtype=int64)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=bfloat16, numpy=array([1, 2], dtype=bfloat16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint16, numpy=array([1, 2], dtype=uint16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float64, numpy=array([1., 2.])>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1., 2.], dtype=float16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint32, numpy=array([1, 2], dtype=uint32)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint64, numpy=array([1, 2], dtype=uint64)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
```
```

Comment (Madjid-CH) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:666]:
```
for d in ('float32', 'float64', 'int32', 'uint8', 'int16', 
          'int8', 'complex64', 'int64', 'bfloat16', 'uint16', 
          'complex128', 'float16', 'uint32', 'uint64'): 
    x = tf.constant([1, 2,], dtype=d)
    print(cummax(x))        
    
(<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 2.], dtype=float32)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float64, numpy=array([1., 2.])>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int32, numpy=array([1, 2])>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint8, numpy=array([1, 2], dtype=uint8)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int16, numpy=array([1, 2], dtype=int16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int8, numpy=array([1, 2], dtype=int8)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 2.], dtype=float32)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=int64, numpy=array([1, 2], dtype=int64)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=bfloat16, numpy=array([1, 2], dtype=bfloat16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint16, numpy=array([1, 2], dtype=uint16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float64, numpy=array([1., 2.])>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1., 2.], dtype=float16)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint32, numpy=array([1, 2], dtype=uint32)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)
(<tf.Tensor: shape=(2,), dtype=uint64, numpy=array([1, 2], dtype=uint64)>, <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1], dtype=int64)>)

```

File: ivy/functional/backends/tensorflow/linear_algebra.py (hunk 5 above)
Comment (Madjid-CH) [reply ivy/functional/backends/tensorflow/linear_algebra.py:591]:
```
for d in ('float32', 'float64', 'int32', 'complex64', 
           'int64', 'bfloat16', 'complex128', 'float16'):
    x1 = tf.constant([1, 2, 3], dtype=d)
    x2 = tf.constant([1, 2, 3], dtype=d)
    print(d, tensordot(x1, x2, axes=0))
    
float32 tf.Tensor(
[[1. 2. 3.]
 [2. 4. 6.]
 [3. 6. 9.]], shape=(3, 3), dtype=float32)
float64 tf.Tensor(
[[1. 2. 3.]
 [2. 4. 6.]
 [3. 6. 9.]], shape=(3, 3), dtype=float64)
int32 tf.Tensor(
[[1 2 3]
 [2 4 6]
 [3 6 9]], shape=(3, 3), dtype=int32)
complex64 tf.Tensor(
[[1.+0.j 2.+0.j 3.+0.j]
 [2.+0.j 4.+0.j 6.+0.j]
 [3.+0.j 6.+0.j 9.+0.j]], shape=(3, 3), dtype=complex64)
int64 tf.Tensor(
[[1 2 3]
 [2 4 6]
 [3 6 9]], shape=(3, 3), dtype=int64)
bfloat16 tf.Tensor(
[[1 2 3]
 [2 4 6]
 [3 6 9]], shape=(3, 3), dtype=bfloat16)
complex128 tf.Tensor(
[[1.+0.j 2.+0.j 3.+0.j]
 [2.+0.j 4.+0.j 6.+0.j]
 [3.+0.j 6.+0.j 9.+0.j]], shape=(3, 3), dtype=complex128)
float16 tf.Tensor(
[[1. 2. 3.]
 [2. 4. 6.]
 [3. 6. 9.]], shape=(3, 3), dtype=float16)

```

File: ivy/functional/backends/tensorflow/elementwise.py (hunk 1 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/tensorflow/elementwise.py:599]: awesome, thanks for clarifying!

File: ivy/functional/backends/tensorflow/elementwise.py (hunk 2 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/tensorflow/elementwise.py:795]: awesome, thanks for clarifying!

File: ivy/functional/backends/tensorflow/elementwise.py (hunk 3 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/tensorflow/elementwise.py:808]: awesome, thanks for clarifying!

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 4 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:666]: awesome, thanks for clarifying!

File: ivy/functional/backends/tensorflow/linear_algebra.py (hunk 5 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/tensorflow/linear_algebra.py:591]: awesome, thanks for clarifying!

File: ivy/functional/backends/tensorflow/linear_algebra.py (hunk 6 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/tensorflow/linear_algebra.py:616]: awesome, thanks for clarifying!

**Review by vedpatwardhan — APPROVED** [review]
lgtm! Feel free to merge, thanks @Madjid-CH 😄
(I assume you've also run the tests for all the functions that are changed and they continue to pass or fail, whatever was the state before making the changes. And even for test failures the logs are the same as before making the changes)

**Pull request merged**